packages/traceability/issuer-api/src/pods/certificate/transaction-log.service.ts

```diff
@@ -0,0 +1,169 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository, In } from 'typeorm';
+import { BigNumber, constants } from 'ethers';
+import { IBlockchainEvent } from '@energyweb/issuer';
+import { validate } from 'class-validator';
+import { NewTransactionProcessedData } from './events/new-transaction-processed.event';
+import { TransactionLog } from './transaction-log.entity';
+import { BlockchainEventType } from './types';
+import { groupBy, partition } from '../../utils/array';
+
+interface ISaveParams {
+    certificateId: number;
+    transactionHash: string;
+    transactionType: BlockchainEventType;
+    transactionTimestamp: Date;
+}
+
+@Injectable()
+export class TransactionLogService {
+    private readonly logger = new Logger(TransactionLogService.name);
+
+    constructor(
+        @InjectRepository(TransactionLog)
+        private readonly repository: Repository<TransactionLog>
+    ) {}
+
+    public async logEvent({ event, transactionType }: NewTransactionProcessedData): Promise<void> {
+        if (this.shouldSkip(event, transactionType)) {
+            return;
+        }
+
+        const certificateIds = this.extractEventCertificateIds(event, transactionType);
+        const savePromises = certificateIds.map((certificateId) =>
+            this.saveTransaction({
+                certificateId,
+                transactionType,
+                transactionHash: event.transactionHash,
+                transactionTimestamp: event.timestamp
+                    ? new Date(event.timestamp * 1000)
+                    : new Date()
+            })
+        );
+
+        await Promise.all(savePromises);
+    }
+
+    public async findByCertificateIds(certificateIds: number[]): Promise<TransactionLog[]> {
+        const logs = await this.repository.find({
+            where: {
+                certificateId: In(certificateIds)
+            }
+        });
+
+        return this.removeDuplicatedLogs(logs);
+    }
+
+    private async saveTransaction({
+        certificateId,
+        transactionHash,
+        transactionType,
+        transactionTimestamp
+    }: ISaveParams): Promise<void> {
+        const entity = this.repository.create({
+            certificateId,
+            transactionHash,
+            transactionType,
+            transactionTimestamp
+        });
+
+        const errors = await validate(entity);
+
+        if (errors.length > 0) {
+            throw new Error(`Transaction save failed: ${errors.join(', ')}`);
```

Comment:
We can also replace `Error` with `UnprocessableEntityException()` from NestJS, to be more explicit. Not a big deal though.